blob/blob.go

```diff
@@ -0,0 +1,801 @@
+	}
+	u := f.containerURL.NewBlobURL(name)
+
+	props, err := u.GetProperties(ctx, azblob.BlobAccessConditions{}, azblob.ClientProvidedKeyOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return newFileInfo(name, props), nil
+}
+
+func (f *FS) dirFile(ctx context.Context, name string) (*File, error) {
+	switch name {
+	case ".", "":
+		return &File{
+			path:    ".",
+			contURL: f.containerURL,
+			fi: fileInfo{
+				name: ".",
+				dir:  true,
+			},
+		}, nil
+	}
+
+	resp, err := f.containerURL.ListBlobsHierarchySegment(
+		ctx,
+		azblob.Marker{},
+		"/",
+		azblob.ListBlobsSegmentOptions{Prefix: name + `/`, MaxResults: math.MaxInt32},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Segment.BlobPrefixes) > 0 || len(resp.Segment.BlobItems) > 0 {
+		return &File{
+			path:    name,
+			contURL: f.containerURL,
+			fi: fileInfo{
+				name: path.Base(name),
+				dir:  true,
+			},
+		}, nil
+	}
+
+	return nil, &fs.PathError{
+		Op:   "open",
+		Path: name,
+		Err:  errors.New("no such file or directory"),
+	}
+}
+
+type rwOptions struct {
+	lock bool
+	tm   azblob.TransferManager
+}
+
+// WithLock locks the file and attempts to keep it locked until the file is closed.
+// If the file in question is a directory, no lease it taken out.
+func WithLock() jsfs.OFOption {
+	return func(o interface{}) error {
+		opt, ok := o.(*rwOptions)
+		if !ok {
+			return fmt.Errorf("WithLock passed to incorrect function")
+		}
+		opt.lock = true
+		return nil
+	}
+}
+
+// WithTransferManager allows you to provide one of azblob's TransferManagers or your
+// own TransferManager for controlling file writes.
+func WithTransferManager(tm azblob.TransferManager) jsfs.OFOption {
+	return func(o interface{}) error {
+		opt, ok := o.(*rwOptions)
+		if !ok {
+			return fmt.Errorf("WithTransferManager passed to incorrect function")
+		}
+		opt.tm = tm
+		return nil
+	}
+}
+
+// These are a subset of what is in "os" that is supported by this file systm.
+const (
+	O_RDONLY int = syscall.O_RDONLY // open the file read-only.
+	O_WRONLY int = syscall.O_WRONLY // open the file write-only.
+	// The remaining values may be or'ed in to control behavior.
+	//O_APPEND int = syscall.O_APPEND // append data to the file when writing.
+	O_CREATE int = syscall.O_CREAT // create a new file if none exists.
+	O_EXCL   int = syscall.O_EXCL  // used with O_CREATE, file must not exist.
+	O_TRUNC  int = syscall.O_TRUNC // truncate regular writable file when opened.
+)
+
+func isFlagSet(flags int, flag int) bool {
+	return flags&flag != 0
+}
+
+// OpenFile implements github.com/johnsiilver/fs.OpenFilerFS. When creating a new file, this will always be a block blob.
+// The fs.File returned will always be a *File.
+func (f *FS) OpenFile(name string, flags int, options ...jsfs.OFOption) (fs.File, error) {
+	opts := rwOptions{}
+	for _, o := range options {
+		if err := o(&opts); err != nil {
+			return nil, err
+		}
+	}
+
+	if opts.lock {
+		if !isFlagSet(flags, O_WRONLY) {
```

Comment:
nit: worth combining these lines together?

```suggestion
	if opts.lock && !isFlagSet(flags, O_WRONLY) {
```

Reply:
Done